### ISSUE TITLE
If a `containerElements` element doesn't exist, focus-trap is confused

### ISSUE THREAD

**Issue by zioth** [focus-trap/focus-trap#223]
If you set `containerElements` to `[null]`, every tabbable node in the document is reachable. I believe this also happens when one of the container elements is no longer in the DOM.

**Comment by stefcameron** [comment]
FYI @cgood92 are you seeing this also?

**Comment by cgood92** [comment]
Well, in this scenario, the trap isn't set yet.  Which seems correct to me.  What would you expect to have happen in this scenario?

**Comment by stefcameron** [comment]
And because of https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L66, the trap won't be created, hence all the nodes are tabbable.

This feels related to refs and possibly what's going on in #183...

**Comment by zioth** [comment]
The `null` case makes sense, but if the node just isn't valid or isn't in the DOM, I feel like the trap should still work. I might have to debug a little more, because I saw a bunch of strange behavior before I got it to work, like:
* Both `null` and a `ref` in the list, and there was no trap.
* Once, the whole document was tabbable _except_ the trapped node.

Unfortunately, I don't remember exactly how I caused these issues.

**Comment by stefcameron** [comment]
How would the trap work without any containers? `null` is not a valid HTMLElement.

**Comment by stefcameron** [comment]
Or you're saying there's some case where the trap is getting an "invalid node" (what would that be?) or one that isn't in the DOM. If it's a valid node but isn't in the DOM, which means it's a new node from `createElement()` that is "unattached", that's an interesting scenario. I'm not sure what happens in that case. I'm thinking the trap doesn't care, but I'm not sure what the results would be. But then, how could you expect an unattached node to work effectively as a trap.

Or is this related to focus-trap-react and refs...?

**Comment by zioth** [comment]
I'll close this issue for now. If I have time later, I'll figure out whether there are some reproducible scenarios other than `containerElements={[null]}`

**Comment by zioth** [comment]
I found the real issue. Steps to reproduce:
* Add two valid nodes to `containerElements`.
* Remove all tabbable elements from one of those nodes (in my case, I have an alert which I close and delete. The parent of the alert is still in the DOM, and it's in `containerElements`).
* Press the tab key.
* Get an exception: "Your focus-trap needs to have at least one focusable element"
* The whole document is tabbable.

It looks like that exception doesn't take the multiple containers into account.

**Comment by stefcameron** [comment]
@zioth Thanks for the details on that! I'll try to have a look soon.

**Comment by stefcameron** [comment]
@zioth Unfortunately, I can't reproduce this as you've described it.

![ft-issue-223](https://user-images.githubusercontent.com/2855350/100268446-ad6a1d80-2f1a-11eb-9fe3-81d2d78f40f0.gif)

See #237 for my interpretation of the steps you described in a new demo on the demo page...

**Comment by zioth** [comment]
Is that recording from code I can look at? Maybe I can modify your test to reproduce the bug.

**Comment by stefcameron** [comment]
Yes, the recording is from the `multipleelements-delete` demo I added in #237. You should be able to get that code into your fork and play with it and post a PR into that branch.

**Comment by zioth** [comment]
This codesandbox reproduces the bug: https://codesandbox.io/s/focus-trapreact-containerelements-demos-forked-7gdn2?file=/src/index.js .

In case that sandbox gets deleted, here's the code:
```
import React, { useState, useCallback } from "react";
import ReactDOM from "react-dom";
import FocusTrap from "focus-trap-react";

const App = () => {
  const [first, setFirst] = useState();
  const [second, setSecond] = useState();
  const [hasTabs, setHasTabs] = useState(true);
  const toggleTabs = useCallback(() => setHasTabs(false));

  return (
    <>
      <div>
        <div>https://github.com/focus-trap/focus-trap/issues/223</div>
        <div style={{ paddingTop: 20 }}>
          Before clicking the button, only sets 1 and 2 are tabbable. After
          clicking it, set 3 is also tabble, along with the rest of the
          document.
        </div>
      </div>
      <div className="first" style={{ paddingTop: 20 }} ref={setFirst}>
        Set 1: {hasTabs && <a href="#">Tabbable node</a>}
      </div>
      <div className="second" style={{ paddingTop: 20 }} ref={setSecond}>
        Set 2: <a href="#">Tabbable node</a>
        <div>
          <button onClick={toggleTabs}>Toggle nodes in set 1</button>
        </div>
      </div>
      <div style={{ paddingTop: 20 }}>
        Set 3: <a href="#">Non-tabbable node</a>
      </div>
      <FocusTrap containerElements={[first, second]}>
        <div />
      </FocusTrap>
    </>
  );
};
```

**Comment by zioth** [comment]
I just noticed something that might point to the cause of this bug: Running the code above, if the first thing I do is press the button, the bug occurs. However, if I first press tab, then the button, the bug doesn't occur.

**Comment by stefcameron** [comment]
Thanks for the sandbox and the tips, I'll check this out!

**Comment by stefcameron** [comment]
The sandbox you provided is using `focus-trap-react`, which adds a level of complexity on top of `focus-trap`. Thankfully, though, if I follow the same "user steps" as you in the test I added to my [branch](https://github.com/focus-trap/focus-trap/pull/237), I can reproduce the issue. So I think it's still a `focus-trap` issue, not a `focus-trap-react` issue.

Some observations with your sandbox:
1. If I immediately use the mouse to click on the button, while the node in the first container still has focus (this seems to be important), then focus gets set out of the trap to the non-tabbable node and the trap no longer works in either direction (tab/shift+tab).
    - I get the same behavior in my branch with my test.
2. If I tab to the node in the second container, such that the first node no longer has focus, then click the button with the mouse, or click it with the enter/spacebar key, the trap continues to work -- but only in the forward direction (tab). If I reserve (shift+tab), then it's like the trap doesn't exist.
    - In my branch, it's similar, but not the same: The continues to work in the forward direction (tab), but when reversing, I can't shift+tab passed the "remove button" button. It doesn't wrap, and it doesn't go anywhere either. It's just stuck there.
3. If I tab to the node in the second container, then click with my mouse on the node in the first container, then click with mouse on the button to remove the node (while it has focus), the trap is busted in either direction.
    - In my branch, this results in the same _similar_ behavior as in (2) where forward works, but reverse gets stuck.

**Comment by stefcameron** [comment]
I have a fix. I'll be updating my branch shortly.

**Comment by stefcameron** [comment]
@zioth Please see #237. It fixes the bug.

**Comment by stefcameron** [comment]
@all-contributors add @zioth for bug.

**Comment by zioth** [comment]
Thanks!

**Comment by stefcameron** [comment]
Fixed in v6.2.2
